README.md
# http-dev-server
A simple http server for develop purpose, support rewrite request, mock response data and serve static resources.


##Install
Install on global path
````shell
npm install -g http-dev-server   
````
Or install as a development dependency
````shell
npm install --save-dev http-dev-server   
````

##Usage
Run server by global command
````shell
http-dev-server -f config-sample.js
````
Or config npm script and run server by npm (for example: {"server": "http-dev-server -f ./dev-server-config.js}")
````shell
npm run server
````

##Config Sample 
the content of config-sample.js
````js
module.exports = {
    hostname: '0.0.0.0',
    port: '3000',
    webPath: __dirname,
    mockPath: __dirname,
    logLevel: 'debug',
    proxies: {
        '/api': {
            host: 'api.examples.com',
        },
        '/userapi': {
            host: '127.0.0.1:7000',
            headers: {
                host: 'api.examples.com'
            }
        }
    },
    mocks: {
        '/api/item/list': 'testdata/test.json',
        '/api/item/1': function(req, res) {
            res.json({ code: 0, data: { id: 1, name: "javascript in action" } });
        },
        '/api/cart': {
            'get': 'testdata/test.json',
            'post': function(req, res) {
                res.send("add cart successful");
            }
        }
    }
};
````

##Test
Run project test using mocha
````shell
npm run test   
````

##Refs
- The function params "req, res" are defined in express: http://expressjs.com/en/4x/api.html
- This repository modified from https://github.com/suxiaoxin/http-server-dev


CHANGES.txt
0.2.2
======
1. fix file upload issue using proxy

0.2.3
======
1. try list files when there are no index.html file in a directory

test/server-test.js
var server = require('../index');
var request = require('supertest');
var config = require('./test-config.js');

describe('server tests:', function() {
    var app = null

    before(function() {
        app = server(config);
    });

    describe('mock::process [GET] request', function() {
        it('should get function response', function(done) {
            request(app)
                .get('/api/msg')
                .expect(200, {msg: 'hello'}, done);
        });

        it('mock::should get the list from file', function(done) {
            request(app)
                .get('/api/list')
                .expect(200)
                .expect(function (res) {
                    var list = res.body || [];
                    if (list.length != 3) {
                        throw new Error("/api/list did not return expected result");
                    }
                })
                .end(done);
        });
    });

    describe('mock::process nested [GET] request', function() {
        it('should get the list from file', function(done) {
            request(app)
                .get('/api/cart')
                .expect(200)
                .expect(function (res) {
                    var list = res.body || [];
                    if (list.length != 3) {
                        throw new Error("/api/list did not return expected result");
                    }
                })
                .end(done);
        });
    });

    describe('mock::process [POST] request', function() {
        it('should add cart successful', function(done) {
            request(app)
                .post('/api/cart')
                .expect(200, {msg: 'successful'}, done);
        });
    });

    describe('proxy::process [GET] request', function() {
        it('should get function response', function(done) {
            request(app)
                .get('/msg')
                .expect(200, {msg: 'hello'}, done);
        });

        it('should add cart successful', function(done) {
            request(app)
                .post('/addcart')
                .expect(200, {msg: 'successful'}, done);
        });
    });


});


test/test-config.js
module.exports = {
    hostname: '127.0.0.1',
    port: '3000',
    webPath: __dirname,
    mockPath: __dirname,
    logLevel: 'error',
    proxies: {
        '/msg': {
            host: '127.0.0.1:3000',
            prefixPath: '/api/msg'
        },

        '/addcart': {
            host: '127.0.0.1:3000',
            prefixPath: '/api/cart',
            headers: {
                host: 'api.examples.com'
            }
        }
    },
    mocks: {
        '/api/list': './test-data.json',
        '/api/msg': function(req, res) {
            res.json( {msg: 'hello' });
        },
        '/api/cart': {
            'get': './test-data.json',
            'post': function(req, res) {
                res.json({msg: 'successful'});
            }
        }
    }
};

demo.js
var server = require('./index');
var config = {
    hostname: '0.0.0.0',
    port: '3000',
    webPath: [__dirname],
    mockPath: __dirname,
    proxies: {
        '/api': {
            prefixPath: '/api',
            host: '127.0.0.1:6000',
        },
        '/userapi': {
            host: '127.0.0.1:7000',
            headers: {
                host: 'api.examples.com'
            }
        }
    },
    mocks: {
        '/api/item/list': 'testdata/test.json',
        '/api/item/1': function(req, res) {
            res.json({ code: 0, data: { id: 1, name: "javascript in action" } });
        },
        '/api/cart': {
            'get': 'testdata/test.json',
            'post': function(req, res) {
                res.send("add cart successful");
            }
        }
    }
};

server(config, function(app) {
    //app is express,  app = express()
    console.log('init...')
})


index.js
var httpProxy = require('express-http-proxy');
var serveIndex = require('serve-index');
var express = require('express');
var _       = require('underscore');
var app = express();
var net = require('net');
var fs  = require('fs');
var logger = require('./lib/logger');


var defaultConfig = {
    hostname : '0.0.0.0',
    port: '3000',
    webPath:  process.cwd(),
    mockPath: process.cwd(),
    logLevel: 'info',
    proxies: {
        /*
        '/api': {
            host: '127.0.0.1:6000',
        },
        '/userapi': {
            host: '127.0.0.1:7000',
            headers: {
                host: 'api.examples.com'
            }
        }*/
    },
    mocks: {
        /*
        '/api/item/list': 'testdata/test.json',
        '/api/item/1': function(req, res) {
            res.json({ code: 0, data: { id: 1, name: 'javascript in action' } });
        },
        '/api/cart': {
            'get': 'testdata/test.json',
            'post': function(req, res) {
                res.send('add cart successful');
            }
        }*/
    }
}

module.exports =  function(config){
    if(!_.isObject(config)){
        logger.error('config is invalid, please check your config.');
        return;
    }

    config = _.defaults(config, defaultConfig);
    //set logger level
    logger.logLevel = config.logLevel || 'info';

    //mocks
    _.each(config.mocks, function (mockConfig, key) {
        if (_.isString(mockConfig)) {
            logger.info("[GET] %s -> %s", key, mockConfig);
                
            app.get(key, function (req, res) {
                logger.debug("[GET] %s -> %s", key, mockConfig);
                res.sendFile(mockConfig, getSendFileOptions(config.mockPath));
            });
        } else if (_.isFunction(mockConfig)) {
            logger.info("[GET] %s -> [function]", key);
            app.get(key, mockConfig);
        } else if (_.isObject(mockConfig)) {
            var route = app.route(key);
            _.each(mockConfig, function (methodConfig, method) {
                if (_.isString(methodConfig)) {
                    logger.info("[%s] %s -> %s", method.toUpperCase(), key, methodConfig);
                    route[method](function (req, res) {
                        logger.debug("[%s] %s -> %s", method.toUpperCase(), key, methodConfig);
                        res.sendFile(methodConfig, getSendFileOptions(config.mockPath));
                    })
                } else if (_.isFunction(methodConfig)) {
                    logger.info("[%s] %s -> [function]", method.toUpperCase(), key);
                    route[method](methodConfig);
                }
            });
        }
    });

    //proxies
    _.each(config.proxies, function(proxyConfig, key){
        var router = express.Router();
        var host = proxyConfig.host;

        if(host.indexOf('http://') != 0 && host.indexOf('https://') != 0){
            host = 'http://' + proxyConfig.host;
        }

        logger.info("[proxy] %s - > %s", key, host);
        
        router.all('*', httpProxy(host, {
            reqBodyEncoding: null,
            decorateRequest: function(req) {
                var path = req.path;

                if (proxyConfig.prefixPath) {
                    req.path = proxyConfig.prefixPath + path;
                }

                logger.debug("[proxy] %s -> %s", key + path, host + req.path);

                if(!_.isEmpty(proxyConfig.headers)){
                    _.each(proxyConfig.headers, function(header, key){
                        req.headers[key] = header;
                    });
                }
                
                return req;
            }
        }))

        app.use(key ,router);
    });

    //static resources
    if (config.webPath) {
        app.use(express.static(config.webPath));
        app.use(serveIndex(config.webPath, {icons: true}));
    }
   
    app.listen(config.port, config.hostname, function(){
        logger.info('Starting http dev server: http://' + config.hostname +':' + config.port);
    });

    return app;
}

function getSendFileOptions(mockPath) {
    return {
        root: mockPath,
        headers: {
            'x-timestamp': Date.now(),
            'x-sent': true
        }
      };
}
